src/test/java/codeu/controller/AdminServletTest.java

```diff
@@ -16,49 +18,454 @@
+
+    adminServlet.doGet(mockRequest, mockResponse);
+
+    Mockito.verify(mockRequest).setAttribute("newestUser", "N/A");
+  }
+
+  @Test
+  public void testNewestUser_MultipleUsers() throws IOException, ServletException {
+    Mockito.when(mockSession.getAttribute("user")).thenReturn("admin");
+
+    User u1 = new User(UUID.randomUUID(), "user1", "hashed_pw", Instant.now());
+    User u2 = new User(UUID.randomUUID(), "user2", "hashed_pw", Instant.now());
+    User u3 = new User(UUID.randomUUID(), "user3", "hashed_pw", Instant.now());
+
+    List<User> mockAllUsers = new ArrayList<User>();
+    mockAllUsers.add(u1);
+    mockAllUsers.add(u2);
+    mockAllUsers.add(u3);
+
+    Mockito.when(mockUserStore.getAllUsers()).thenReturn(mockAllUsers);
+
+    adminServlet.doGet(mockRequest, mockResponse);
+
+    Mockito.verify(mockRequest).setAttribute("newestUser", "user3");
+  }
+
+  // Most active user tests
+  @Test
+  public void testMostActiveUser_NoUsers() throws IOException, ServletException {
+    Mockito.when(mockSession.getAttribute("user")).thenReturn("admin");
 
-    @Test
-    public void testDoGet_Admin() throws IOException, ServletException {
-        Mockito.when(mockSession.getAttribute("user")).thenReturn("daniel");
+    List<User> mockAllUsers = new ArrayList<User>();
+    Mockito.when(mockUserStore.getAllUsers()).thenReturn(mockAllUsers);
 
-        adminServlet.doGet(mockRequest, mockResponse);
+    adminServlet.doGet(mockRequest, mockResponse);
 
-        Mockito.verify(mockRequestDispatcher).forward(mockRequest, mockResponse);
+    Mockito.verify(mockRequest).setAttribute("mostActiveUser", "N/A");
+  }
+
+  @Test
+  public void testMostActiveUser_MultipleUsers() throws IOException, ServletException {
+    Mockito.when(mockSession.getAttribute("user")).thenReturn("admin");
+
+    // Set up mock users
+    User u1 = new User(UUID.randomUUID(), "user1", "hashed_pw", Instant.now());
+    User u2 = new User(UUID.randomUUID(), "user2", "hashed_pw", Instant.now());
+    User u3 = new User(UUID.randomUUID(), "user3", "hashed_pw", Instant.now());
+
+    List<User> mockAllUsers = new ArrayList<User>();
+    mockAllUsers.add(u1);
+    mockAllUsers.add(u2);
+    mockAllUsers.add(u3);
+
+    Mockito.when(mockUserStore.getAllUsers()).thenReturn(mockAllUsers);
+
+
+    // Set up mock conversations
+    Conversation mockConversation = new Conversation(UUID.randomUUID(), u1.getId(), "mock conversation", Instant.now());
+
+    List<Conversation> mockAllConversations = new ArrayList<Conversation>();
+    mockAllConversations.add(mockConversation);
+
+    Mockito.when(mockConversationStore.getAllConversations()).thenReturn(mockAllConversations);
+
+
+    // Add mock messages
+    List<Message> mockMessagesInConversation = new ArrayList<Message>();
+    String[][] messages = {
+      {"a"},
+      {"a", "b"},
+      {"a", "b", "c"}
+    };
+    User[] users = {u1, u2, u3};
+
+    for (int u = 0; u < messages.length; u++) {
```

Comment:
Hm...since it's important for the test that user3 have the most messages, it might be clearer if we add the conversations in a less nested way. It's probably also fine for this test if all of the message contents are the same, so that might make it a little simpler to set up the test data.